packages/react-dom/src/server/ReactDOMServerFormatConfig.js

```diff
@@ -1128,12 +1128,17 @@ function pushStartCustomElement(
 
   let children = null;
   let innerHTML = null;
-  for (const propKey in props) {
+  for (let propKey in props) {
     if (hasOwnProperty.call(props, propKey)) {
       const propValue = props[propKey];
       if (propValue == null) {
         continue;
       }
+      if (enableCustomElementPropertySupport && propKey === 'className') {
+        // className gets rendered as class on the client, so it should be
+        // rendered as class on the server.
+        propKey = 'class';
```

Comment:
className is special because all elements have a builtin `className` property which triggers the `'className' in node` check when rendering on the client - meaning that JSX on the client will always use a property instead of an attribute for className. Without this, one of the automated tests was getting errors (I think during hydration?) due to this mismatch.
Alternatively we could make it so that when client rendering `className` is always set as an attribute instead of a property, but I think this makes more sense and it's what Preact is already doing.

Reply:
Yea this make sense because htmlFor doesn't have any semantics on custom element which are just Generic HTML elements.

There are a number of other ones that are case sensitive for similar reasons. contentEditable, tabIndex, etc. They probably work if used correctly, but it would be nice to add a warning if they use the wrong case.

There are some that just won't work like `offsetHeight` etc. Since those exist as properties but are read-only. Probably should ensure we have a nice warning about those and why they don't work. So that if they're rendered on the server, it doesn't look like they work.

Similarly there are some we have to be careful with like `innerHTML` and `innerText`. It shouldn't pass through as a property to custom elements - even in production so it can't be a DEV only error. Could be worth a test.

Reply:
I'm surprised innerText and textContent aren't reserved properties in React. Doesn't seem right.

Reply:
> There are a number of other ones that are case sensitive for similar reasons. contentEditable, tabIndex, etc. They probably work if used correctly, but it would be nice to add a warning if they use the wrong case.

contentEditable is interesting, it looks like the attribute can be either `contentEditable` or `contenteditable`, they both work, and the reflected object property is `contentEditable`.
So I guess if someone writes `<my-custom-element contentEditable={true} />` or `<my-custom-element contenteditable={true} />` they should both work fine, right...?

> There are some that just won't work like offsetHeight etc. Since those exist as properties but are read-only. Probably should ensure we have a nice warning about those and why they don't work. So that if they're rendered on the server, it doesn't look like they work.

I mean, you *can* assign to offsetHeight from script and the browser doesn't complain about it... but yeah I guess it would be weird to have a state where you can assign to an attribute called `offsetHeight` when rendering on the server, but you can't on the client because it will always trigger the `'offsetHeight' in node` check...

Should I consider every global attribute and every built in property?

Reply:
From offline:
- for camel case reflected attribute/properties like `contentEditable`, we should just have a warning if you server render it as `contenteditable` instead of `contentEditable`. The actual behavior will likely be identical if you choose `contentEditable` or `contenteditable` because in the lowercase form it will assign to the attribute, get reflected to the property, and work - and if you do `contentEditable`, then it will assign to the property, reflect to the attribute, and still work just fine.
- for `innerHTML`, `innerText`, and `textContent`, we should probably just block them like we do for regular elements. If a custom element tries to overload these and needs them in order to function, then maybe that case wouldn't work but I'm not sure it's a very common use case.
  - Sebastian mentioned that there is a list somewhere of blocked properties which has innerHTML and that we should add textContent to it separately from this PR.
- for read only properties like `offsetHeight`, we can just have a dev mode warning. When assigning to these in the browser, it's just a no-op that doesn't throw an error or anything, and it doesn't affect the return value of the property (at least for `offsetHeight`), so we shouldn't be worried about assigning to them.

Reply:
Re: blocklist.

https://github.com/facebook/react/blob/50263d3273b6fc983acc5b0fd52e670399b248b1/packages/react-dom/src/shared/DOMProperty.js#L236-L249

Reply:
Thanks Dan! I went ahead and added innerText and textContent to that list in this PR behind the flag

Reply:
I've thought some more about `contentEditable` vs `contenteditable`, and I'm not sure if we should emit any warnings...

I went through all of the global reflected HTML attributes and made a table at the bottom to show their reflected object property name vs the react one, and I found that there are some like `contentEditable` where the React name and the reflected object property are the same, which would result in client rendering passing the `in` check and using the object property, and then there are some like `autofocus`/`autoFocus` where using the React name of `autoFocus` would cause the `in` check to fail and we would end up using the attribute instead...

Basically I don't see the point of a warning based on capitalization because:
1. However you capitalize the JSX attribute its going to end up the same - we don't yet have any evidence to suggest that doing `element.contentEditable = true` is any different from `element.setAttribute('contenteditable', 'true')`
2. There are discrepancies between the React name capitalization and the reflected object property name capitalization - why should we tell developers that they should use the React style capitalization, the object property name capitalization, or the HTML attribute capitalization for custom elements if we aren't doing so already?

HTML attribute (always lowercase) | object property | React JSX attribute name
----|----|----
`class`|`className`|`className`
`contenteditable`|`contentEditable`|`contentEditable`
`accesskey`|`accessKey`|`accessKey`
`autofocus`|`autofocus`|`autoFocus`
`dir`|`dir`|`dir`
`draggable`|`draggable`|`draggable`
`enterkeyhint`|`enterKeyHint`|`enterKeyHint`
`hidden`|`hidden`|`hidden`
`id`|`id`|`id`
`inputmode`|`inputMode`|`inputMode`
`lang`|`lang`|`lang`
`nonce`|`nonce`|`nonce`
`part`|`part`|`part`
`slot`|`slot`|`slot`
`spellcheck`|`spellcheck`|`spellCheck`
`style`|`style`|`style`
`tabindex`|`tabIndex`|`tabIndex`
`title`|`title`|`title`
`translate`|`translate`|`translate`

Writing this table also made me realize something about Dan's initial question about `htmlFor` - `htmlFor` isn't relevant for the special casing for `className` here because custom elements won't get a `htmlFor` object property set by the browser - only built in form control elements have those.

Reply:
For the read only properties like offsetHeight, I just pushed a commit which emits the warning: https://github.com/facebook/react/pull/22184/commits/1a093e584ce50e2e634aa743e04f9cb8fc2b3f7d

However, the test fails because it throws an exception when assigning to the read only properties - but this exception doesn't occur in the browser or in JSDOM... what kind of browser/DOM setup is being used in these tests?

Reply:
“use strict” throws, right?

Reply:
Oh wow, today I learned...
What do you think we should do? Remove the test? Make it somehow get rid of the 'use strict'?

Reply:
I believe if you render it on the server and hydrate it wouldn’t error.

The reason for the warning would be so that if you render it on the server, during development you should know not to.

So you can test hydrating it.

Reply:
Ah, that makes sense!
I replaced the client rendering check with a hydration check and made a test for it.